url56b1dc76d3_wufunxiang_Libs_tgz-pJ8-org_msgpack_util_TemplatePrecompilerJ8/src/org/msgpack/template/builder/ReflectionTemplateBuilder.java

```diff
@@ -0,0 +1,204 @@
+
+    private static Logger LOG = Logger.getLogger(ReflectionBeansTemplateBuilder.class.getName());
+
+    protected static abstract class ReflectionFieldTemplate extends AbstractTemplate<Object> {
+        protected FieldEntry entry;
+
+        ReflectionFieldTemplate(final FieldEntry entry) {
+            this.entry = entry;
+        }
+
+        void setNil(Object v) {
+            entry.set(v, null);
+        }
+    }
+
+    static final class FieldTemplateImpl extends ReflectionFieldTemplate {
+        private Template template;
+
+        public FieldTemplateImpl(final FieldEntry entry, final Template template) {
+            super(entry);
+            this.template = template;
+        }
+
+        
+        public void write(Packer packer, Object v, boolean required)
+                throws IOException {
+            template.write(packer, v, required);
+        }
+
+        
+        public Object read(Unpacker unpacker, Object to, boolean required)
+                throws IOException {
+            // Class<Object> type = (Class<Object>) entry.getType();
+            Object f = entry.get(to);
+            Object o = template.read(unpacker, f, required);
+            if (o != f) {
+                entry.set(to, o);
+            }
+            return o;
+        }
+    }
+
+    protected static class ReflectionClassTemplate<T> extends AbstractTemplate<T> {
+        protected Class<T> targetClass;
+
+        protected ReflectionFieldTemplate[] templates;
+
+        protected ReflectionClassTemplate(Class<T> targetClass, ReflectionFieldTemplate[] templates) {
+            this.targetClass = targetClass;
+            this.templates = templates;
+        }
+
+        
+        public void write(Packer packer, T target, boolean required)
+                throws IOException {
+            if (target == null) {
+                if (required) {
+                    throw new MessageTypeException("attempted to write null");
+                }
+                packer.writeNil();
+                return;
+            }
+            try {
+                packer.writeArrayBegin(templates.length);
+                for (ReflectionFieldTemplate tmpl : templates) {
+                    if (!tmpl.entry.isAvailable()) {
+                        packer.writeNil();
+                        continue;
+                    }
+                    Object obj = tmpl.entry.get(target);
+                    if (obj == null) {
+                        if (tmpl.entry.isNotNullable()) {
+                            throw new MessageTypeException(tmpl.entry.getName()
+                                    + " cannot be null by @NotNullable");
+                        }
+                        packer.writeNil();
+                    } else {
+                        tmpl.write(packer, obj, true);
+                    }
+                }
+                packer.writeArrayEnd();
+            } catch (IOException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An exception is caught and re-thrown within the catch clause without any other operations being performed on the exception, making the catch clause unnecessary.

**Fix**: Perform additional operations on the exception that has been caught. You can wrap a standard exception into a custom exception to add additional information or special handling. When you do that, make sure to set the original exception as the cause. Otherwise, you will lose the stack trace and message of the original exception, which will make it difficult to analyze the event that caused the exception.  
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)